<a href="https://colab.research.google.com/github/boscolio/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [ ]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student
!unzip student.zip

--2020-08-11 17:53:39--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.1’

student.1           100%[===================>]  20.00K  --.-KB/s    in 0.04s   

2020-08-11 17:53:39 (558 KB/s) - ‘student.1’ saved [20478/20478]

Archive:  student.zip
replace student-mat.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student-mat.csv         
replace student-por.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student-por.csv         
replace student-merge.R? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student-merge.R         
replace student.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student.txt             


In [ ]:
# Importing libraries
import pandas as pd
import numpy as np
from scipy import stats

In [ ]:
# This isn't working for me either, just like Ryan.
!curl student-mat.csv

curl: (6) Could not resolve host: student-mat.csv


In [ ]:
# The file is separated by ';' instead of ','
df = pd.read_csv('student-mat.csv', sep=";")
df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


###Test No. 1

1) Null Hypothesis:

Failures and health are independent

2) Alternative Hypothesis:

Failures and health are dependent

In [ ]:
contingency = pd.crosstab(df['failures'], df['health'])

contingency

health     1   2   3   4    5
failures                     
0         42  37  69  54  110
1          3   5  12   6   24
2          1   2   5   2    7
3          1   1   5   4    5

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 8.224511251466813
p value: 0.7673486513090713
dof: 12
expected_frequencies: 
 [[ 37.12405063  35.5443038   71.87848101  52.13164557 115.32151899]
 [  5.94936709   5.69620253  11.51898734   8.35443038  18.48101266]
 [  2.02278481   1.93670886   3.9164557    2.84050633   6.2835443 ]
 [  1.90379747   1.82278481   3.68607595   2.67341772   5.91392405]]


Conclusion:

Based on a chi2 statistic of 8.22 and a p-value of .77 we fail to reject the null hypothesis that failures and health are independant.

###Test No. 2

1) Null Hypothesis:

Failures and travel time are independent

2) Alternative Hypothesis:

Failures and travel time are dependent

In [ ]:
contingency2 = pd.crosstab(df['failures'], df['traveltime'])

contingency2

traveltime    1   2   3  4
failures                  
0           209  80  17  6
1            30  17   3  0
2             7   9   1  0
3            11   1   2  2

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency2)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 21.124872298979817
p value: 0.012107363244231761
dof: 9
expected_frequencies: 
 [[202.99746835  84.5164557   18.16708861   6.31898734]
 [ 32.53164557  13.5443038    2.91139241   1.01265823]
 [ 11.06075949   4.60506329   0.98987342   0.3443038 ]
 [ 10.41012658   4.33417722   0.93164557   0.32405063]]


Conclusion:

Based on a chi2 statistic of 21.12 and a p-value of .012 we reject the null hypothesis that failures and travel time are independant.

###Test No. 3

1) Null Hypothesis:

Travel time and free time are independent

2) Alternative Hypothesis:

Travel time and free time are dependent

In [ ]:
contingency3 = pd.crosstab(df['traveltime'], df['freetime'])

contingency3

freetime     1   2    3   4   5
traveltime                     
1           11  40  106  72  28
2            7  18   38  34  10
3            1   6    8   8   0
4            0   0    5   1   2

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency3)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 11.46774269441963
p value: 0.4893142424435174
dof: 12
expected_frequencies: 
 [[ 12.36202532  41.64050633 102.14936709  74.82278481  26.02531646]
 [  5.14683544  17.33670886  42.52911392  31.15189873  10.83544304]
 [  1.10632911   3.72658228   9.14177215   6.69620253   2.32911392]
 [  0.38481013   1.29620253   3.17974684   2.32911392   0.81012658]]


Conclusion:

Based on a chi2 statistic of 11.47 and a p-value of .49 we fail to reject the null hypothesis that travel time and free time are independant.

###Test No. 4

1) Null Hypothesis:

Free time and study time are independent

2) Alternative Hypothesis:

Free time and study time are dependent

In [ ]:
contingency4 = pd.crosstab(df['freetime'], df['studytime'])

contingency4

studytime   1   2   3   4
freetime                 
1           2  11   3   3
2          19  28  11   6
3          35  82  29  11
4          29  61  21   4
5          20  16   1   3

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency4)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 23.53045426913059
p value: 0.023545617681757027
dof: 12
expected_frequencies: 
 [[ 5.05063291  9.52405063  3.12658228  1.29873418]
 [17.01265823 32.08101266 10.53164557  4.37468354]
 [41.73417722 78.69873418 25.83544304 10.73164557]
 [30.56962025 57.64556962 18.92405063  7.86075949]
 [10.63291139 20.05063291  6.58227848  2.73417722]]


Conclusion:

Based on a chi2 of 23.53 and a p-value of .024 we reject the null hypothesis that free time and study time are independant.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# Creating the dataframe for "free time" and "study time", including the margins.
observed = pd.crosstab(df['freetime'], df['studytime'], margins=-True)

observed

studytime    1    2   3   4  All
freetime                        
1            2   11   3   3   19
2           19   28  11   6   64
3           35   82  29  11  157
4           29   61  21   4  115
5           20   16   1   3   40
All        105  198  65  27  395

In [ ]:
# Using .values to insert the row and column totals into numpy arrays.
# Adding .sort_index() to both row_totals and column_totals was needed.

row_totals = df['freetime'].value_counts().sort_index().values

row_totals

array([ 19,  64, 157, 115,  40])

In [ ]:
column_totals = df['studytime'].value_counts().sort_index().values

column_totals

array([105, 198,  65,  27])

In [ ]:
# Sample size of our data
sample_size = df.shape[0]

sample_size

395

In [ ]:
# Building array with 0.0 as placeholder floats

expected = np.array([[0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0]])

# Using nested for loops to fill the array 
for row_index, row in enumerate(row_totals):
  for column_index, col in enumerate(column_totals):
    # Calculates the expected value
    exp_value = (row*col)/sample_size
    # Inserts that value into 'expected'
    expected[row_index][column_index] = exp_value
    # Prints expected value at the row/column in 'expected'

expected

array([[ 5.05063291,  9.52405063,  3.12658228,  1.29873418],
       [17.01265823, 32.08101266, 10.53164557,  4.37468354],
       [41.73417722, 78.69873418, 25.83544304, 10.73164557],
       [30.56962025, 57.64556962, 18.92405063,  7.86075949],
       [10.63291139, 20.05063291,  6.58227848,  2.73417722]])

In [ ]:
observed = contingency4.values
observed

array([[ 2, 11,  3,  3],
       [19, 28, 11,  6],
       [35, 82, 29, 11],
       [29, 61, 21,  4],
       [20, 16,  1,  3]])

In [ ]:
# Calculating the chi2 statistic
chi2 = (((observed - expected)**2)/ expected).sum()

print('Chi2:', chi2)

Chi2: 23.53045426913059


Inserting this chi2 into the applet we get a p-value of 0.02355

In [ ]:
# Degree of Freedom. Showing the math.
deg_of_free = (5-1)*(4-1)
print('Degree of Freedom:', deg_of_free)

Degree of Freedom: 12


###Conclusion

We were able to replicate and confirm the four values from the `chi2_contingency()` test.

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)